src/EditorFeatures/CSharp/InlineRename/CSharpEditorInlineRenameService.cs

```diff
@@ -18,4 +26,105 @@ internal sealed class CSharpEditorInlineRenameService(
     [ImportMany] IEnumerable<IRefactorNotifyService> refactorNotifyServices,
     IGlobalOptionService globalOptions) : AbstractEditorInlineRenameService(refactorNotifyServices, globalOptions)
 {
+    protected override async Task<ImmutableDictionary<string, ImmutableArray<string>>> GetRenameContextCoreAsync(IInlineRenameInfo inlineRenameInfo, IInlineRenameLocationSet inlineRenameLocationSet, CancellationToken cancellationToken)
+    {
+        var seen = PooledHashSet<TextSpan>.GetInstance();
+        var definitions = ArrayBuilder<string>.GetInstance();
+        var references = ArrayBuilder<string>.GetInstance();
+        var docComments = ArrayBuilder<string>.GetInstance();
+
+        foreach (var renameDefinition in inlineRenameInfo.DefinitionLocations)
+        {
+            var containingStatementOrDeclarationSpan =
+                await renameDefinition.Document.TryGetSurroundingNodeSpanAsync<StatementSyntax>(renameDefinition.SourceSpan, cancellationToken).ConfigureAwait(false) ??
+                await renameDefinition.Document.TryGetSurroundingNodeSpanAsync<MemberDeclarationSyntax>(renameDefinition.SourceSpan, cancellationToken).ConfigureAwait(false);
+
+            var documentText = await renameDefinition.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (documentText is null)
+            {
+                continue;
+            }
+
+            var symbolService = renameDefinition.Document.GetLanguageService<IGoToDefinitionSymbolService>();
+            if (symbolService is not null)
+            {
+                var textSpan = inlineRenameInfo.TriggerSpan;
+                var (symbol, _, _) = await symbolService.GetSymbolProjectAndBoundSpanAsync(
+                    renameDefinition.Document, textSpan.Start, cancellationToken)
+                    .ConfigureAwait(true);
+                var docComment = symbol?.GetDocumentationCommentXml(expandIncludes: true, cancellationToken: cancellationToken);
+                if (!string.IsNullOrWhiteSpace(docComment))
+                {
+                    docComments.Add(docComment!);
+                }
+            }
+
+            AddSpanOfInterest(documentText, renameDefinition.SourceSpan, containingStatementOrDeclarationSpan, definitions);
+        }
+
+        foreach (var renameLocation in inlineRenameLocationSet.Locations)
+        {
+            var containingStatementOrDeclarationSpan =
+                await renameLocation.Document.TryGetSurroundingNodeSpanAsync<BaseMethodDeclarationSyntax>(renameLocation.TextSpan, cancellationToken).ConfigureAwait(false) ??
```

Comment:
this needs docs.  i don't understand the order.  first we try the surrounding method, *then* the surrounding statement?  When would the latter hhappen, but not the former?

Reply:
this comment still applies.  

also, you may be getting a *lot* of spans while doing the rename.  is this whole system properly async/non-blocking?  in other words, can the person still perform the rename and everything while this computatio is going on?

Reply:
> also, you may be getting a lot of spans while doing the rename. is this whole system properly async/non-blocking? in other words, can the person still perform the rename and everything while this computatio is going on?

This operation is not blocking the normal rename flow, it's only blocking the copilot result. So if you don't want the AI sugguestion you can just do normal rename.